zebrad/tests/common/cached_state.rs

```diff
@@ -144,3 +158,149 @@ async fn copy_directory(
+        .await?
+        .into_iter()
+        .map(hex::decode)
+        .map(|block_bytes| {
+            block_bytes
+                .expect("getblock rpc calls in get_raw_future_blocks should return valid hexdata")
+                .zcash_deserialize_into()
+                .expect("decoded hex data from getblock rpc calls should deserialize into blocks")
+        })
+        .collect();
+
+    Ok(blocks)
+}
+
+/// Accepts a network, test_type, test_name, and num_blocks (how many blocks past the finalized tip to try getting)
+///
+/// Syncs zebra until the tip, gets some blocks near the tip, via getblock rpc calls,
+/// shuts down zebra, and gets the finalized tip height of the updated cached state.
+///
+/// Returns hexdata of retrieved blocks that are above the finalized tip height of the cached state.
+///
+/// ## Panics
+///
+/// If the provided `test_type` doesn't need an rpc server and cached state, or if `max_num_blocks` is 0
+pub async fn get_raw_future_blocks(
+    network: Network,
+    test_type: TestType,
+    test_name: &str,
+    max_num_blocks: u32,
+) -> Result<Vec<String>> {
+    assert!(max_num_blocks > 0);
+
+    let max_num_blocks = max_num_blocks.min(MAX_BLOCK_REORG_HEIGHT);
+    let mut raw_blocks = Vec::with_capacity(max_num_blocks as usize);
+
+    assert!(
+        test_type.needs_zebra_cached_state() && test_type.needs_zebra_rpc_server(),
+        "get_raw_future_blocks needs zebra cached state and rpc server"
+    );
+
+    let should_sync = true;
+    let (zebrad, zebra_rpc_address) =
+        spawn_zebrad_for_rpc(network, test_name, test_type, should_sync)?
+            .ok_or_else(|| eyre!("get_raw_future_blocks requires a cached state"))?;
+    let rpc_address = zebra_rpc_address.expect("test type must have RPC port");
+
+    let mut zebrad = check_sync_logs_until(
+        zebrad,
+        network,
+        SYNC_FINISHED_REGEX,
+        MempoolBehavior::ShouldAutomaticallyActivate,
+        true,
+    )?;
+
+    // Create an http client
+    let client = Client::new();
+
+    let send_rpc_request = |method, params| {
+        client
+            .post(format!("http://{}", &rpc_address))
+            .body(format!(
+                r#"{{"jsonrpc": "2.0", "method": "{method}", "params": {params}, "id":123 }}"#
+            ))
+            .header("Content-Type", "application/json")
+            .send()
+    };
+
+    let blockchain_info: serde_json::Value = serde_json::from_str(
+        &send_rpc_request("getblockchaininfo", "[]".to_string())
+            .await?
+            .text()
+            .await?,
+    )?;
+
+    let tip_height: u32 = blockchain_info["result"]["blocks"]
+        .as_u64()
+        .expect("unexpected block height: doesn't fit in u64")
+        .try_into()
+        .expect("unexpected block height: doesn't fit in u32");
+
+    let estimated_finalized_tip_height = tip_height - MAX_BLOCK_REORG_HEIGHT;
+
+    tracing::info!(
+        ?estimated_finalized_tip_height,
```

Comment:
I think this might help with debugging:

```suggestion
        ?tip_height,
        ?estimated_finalized_tip_height,
```